File: application/samples/CSALT/ochPlotter.m
clc;clear all;close all;format long g;

% *.och filename
fname = 'DROtoL2.och';

% Get data
[nx, nu, data] = getOchData(fname);

% Extract Data
time = data(:,1);
state = data(:,2:1+nx);
control = data(:,1+nx+1:1+nx+nu);

figure(1)
hold on
plot(state(:,1),state(:,2),'linewidth',2)
quiver(state(:,1),state(:,2),state(:,4),state(:,5))
hold off
axis equal
xlabel('X [ND]')
ylabel('Y [ND]')
grid on
set(gca,'fontsize',14)

% % Plot states
% for i = 1:nx
%     figure(i)
%     plot(time,state(:,i))
%     xlabel('time')
%     ylabel(sprintf('state %d',i))
%     grid on
% end
% 
% % Plot controls
% for i = 1:nu
%     figure(nx+i)
%     plot(time,control(:,i))
%     xlabel('time')
%     ylabel(sprintf('control %d',i))
%     grid on
% end





function [nx, nu, data] = getOchData(fname)
fid = fopen(fname,'r');
indata = false;
data = [];
while ~feof(fid)
    line = split(fgetl(fid));
    if any(strcmp(line,'NUM_STATES'))
        nx = str2double(line{4});
    end
    if any(strcmp(line,'NUM_CONTROLS'))
        nu = str2double(line{4});
    end
    if any(strcmp(line,'DATA_START'))
        indata = true;
        n = 1+nx+nu;
        continue;
    end
    if any(strcmp(line,'DATA_STOP'))
        indata = false;
    end
    if indata
        ldata = zeros(1,n);
        j = 1;
        for i = 1:length(line)
            if ~isempty(line{i})
                ldata(j) = str2double(line{i});
                j = j+1;
            end
        end
        data = [data;ldata];
    end
end
fclose(fid);
end

File: application/samples/CSALT/ochPlotter.py
import os
import matplotlib.pyplot as plt

def getOchData(fname):
    if not os.path.isfile(fname):
        raise Exception("*.och file %s not found" %(fname))
    
    data = []
    indata = False
    with open(fname,'r') as fid:
        for line in fid:
            if "NUM_STATES" in line:
                nx = int(line.split('=')[-1])
            if "NUM_CONTROLS" in line:
                nu = int(line.split('=')[-1])
            if "DATA_START" in line:
                indata = True
                continue
            if "DATA_STOP" in line:
                indata = False
                continue
            if indata is True:
                lsplt = line.split()
                data.append([float(i) for i in lsplt])
    
    return nx, nu, list(zip(*data))

if __name__ == '__main__':
    # *.ocd filename
    fname = 'OrbitRaisingMultiPhase.och'
    
    # Get Data
    nx, nu, data = getOchData(fname)
    
    # Extract Data
    time = data[0]
    state = data[1:1+nx]
    control = data[1+nx:1+nx+nu]
    
    # Plot states
    for i in range(nx):
        plt.figure(i+1)
        plt.plot(time,state[i])
        plt.xlabel('time')
        plt.ylabel('state %s' %(i+1))
        plt.grid()
       
    # Plot controls
    for i in range(nu):
        plt.figure(i+nx+1)
        plt.plot(time,control[i])
        plt.xlabel('time')
        plt.ylabel('control %s' %(i+1))
        plt.grid()
    
    # Deploy plots
    plt.show()